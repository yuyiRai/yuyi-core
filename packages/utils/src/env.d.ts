import "@yuyi919/ts-transformer-awesome/env"
