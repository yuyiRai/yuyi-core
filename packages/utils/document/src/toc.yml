items:
  -
    name: '@yuyi919/utils'
    href: ~/overview/sharepoint.md
    items:
      -
        name: CustomUtils
        uid: CustomUtils!
        items:
          -
            name: 'App (Namespace)'
      -
        name: EventEmitter
        uid: EventEmitter!
        items:
          -
            name: EventEmitter
            uid: 'EventEmitter!EventEmitter:class'
      -
        name: index
        uid: index!
        items:
          -
            name: 'App (Namespace)'
          -
            name: CommonDto
            uid: 'index!CommonDto:class'
          -
            name: 'DateUtils (Class)'
            uid: 'index!DateUtils:class'
          -
            name: 'DateUtils (Namespace)'
            items:
              - { name: EFormatter, uid: 'index!DateUtils.EFormatter:enum' }
          -
            name: EventEmitter
            uid: 'index!EventEmitter:class'
          -
            name: IGetPathResult
            uid: 'index!IGetPathResult:interface'
          -
            name: 'isEqual (Namespace)'
          -
            name: ITypeFilterUtils
            uid: 'index!ITypeFilterUtils:interface'
          -
            name: ITypeUtils
            uid: 'index!ITypeUtils:interface'
          -
            name: Option
            uid: 'index!Option:interface'
      -
        name: LodashExtra
        uid: LodashExtra!
        items:
          -
            name: CustomizerAdapterType
            uid: 'LodashExtra!CustomizerAdapterType:enum'
          -
            name: 'isEqual (Namespace)'
      -
        name: NodeUtils
        uid: NodeUtils!
        items:
          -
            name: Args
            uid: 'NodeUtils!Args:class'
          -
            name: ArgumentOptionsFixed
            uid: 'NodeUtils!ArgumentOptionsFixed:interface'
          -
            name: ArgumentParserFixed
            uid: 'NodeUtils!ArgumentParserFixed:interface'
      -
        name: OptionsUtils
        uid: OptionsUtils!
      -
        name: PropertyUtils
        uid: PropertyUtils!
        items:
          -
            name: CommonDto
            uid: 'PropertyUtils!CommonDto:class'
          -
            name: IGetPathResult
            uid: 'PropertyUtils!IGetPathResult:interface'
      -
        name: TestUtil
        uid: TestUtil!
      -
        name: TestUtils
        uid: TestUtils!
        items:
          -
            name: BenchmarkUtils
      -
        name: TsUtils
        uid: TsUtils!
        items:
          -
            name: Base
            uid: 'TsUtils!Base:class'
          -
            name: IFunction
            uid: 'TsUtils!IFunction:interface'
          -
            name: IKeyValueMap
            uid: 'TsUtils!IKeyValueMap:interface'
      -
        name: TypeLib
        uid: TypeLib!
        items:
          -
            name: ITypeFilterUtils
            uid: 'TypeLib!ITypeFilterUtils:interface'
          -
            name: ITypeUtils
            uid: 'TypeLib!ITypeUtils:interface'
      -
        name: WasmLoader
        uid: WasmLoader!
