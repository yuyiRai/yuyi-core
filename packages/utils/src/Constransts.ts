// tslint:disable: variable-name

declare global {
  namespace NodeJS {
    // @ts-ignore
    interface Global {
      __DEV__: boolean;
    }
  }
}

export type LoopIterator<T, K extends TKey, TResult = void> = (currentValue: T, currentIndex: K, array: T[]) => TResult
export type ArrayIterator<T, TResult = void> = LoopIterator<T, number, TResult>

export namespace FunctionFactory {
  export type Base<Args extends any[] = any[], Result = any> = (...args: Args) => Result;

  export type ReturnType<F, Default = any> = F extends Base<any, infer R> ? R : Default;

  export type Bind1<A, Args extends any[] = [], Result = any> = (a: A, ...args: Args) => Result;
  export type Bind2<A, B, Args extends any[] = [], Result = any> = (a: A, b: B, ...args: Args) => Result;
  export type Bind3<A, B, C, Args extends any[] = [], Result = any> = (a: A, b: B, c: C, ...args: Args) => Result;

  export type Arg0<T> = T extends Bind1<infer T, any[]> ? T : any;
  export type Arg1<T> = T extends Bind2<any, infer T, any[]> ? T : any;
  export type Arg2<T> = T extends Bind3<any, any, infer T, any[]> ? T : any;
  export type ExtractArgs<T, Default = []> = T extends (...args: infer Args) => any ? Args : Default;

  export type Shift1<T, A> = T extends Bind1<A, infer Args, infer Re> ? Base<Args, Re> : Function
  export type Shift2<T, A, B> = T extends Bind2<A, B, infer Args, infer Re> ? Base<Args, Re> : Function
  export type Shift3<T, A, B, C> = T extends Bind3<A, B, C, infer Args, infer Re> ? Base<Args, Re> : Function


  // export type Bind<BindArgs extends any[], Args extends any[] = [], Result = any> = ((binds: BindArgs, ...args: Args) => Result);
  // export interface ShiftAuto<T, A = any, B = any, C = any, Re = ReturnType<T>> {
  //   <Args extends T extends Bind<[A], infer Args> ? Args : any[]>(...args: Args): Re
  //   <Args extends T extends Bind<[A, B], infer Args> ? Args : any[]>(...args: Args): Re
  //   <Args extends T extends Bind<[A, B, C], infer Args> ? Args : any[]>(...args: Args): Re
  // }
  // export type Shift<T, A extends any[]> = T extends Bind<A, infer Args, infer Re> ? Base<Args, Re> : Function
}
// var global = window;

export namespace Constant$ {

  export const ENTRIES = Object.entries
  /**
   * * keyof PropertyDescriptor
   * writable, configurable, value, enumerable
   */
  export enum DefPropDec$$ {
    W = 'writable',
    C = 'configurable',
    V = 'value',
    E = 'enumerable',
    G = 'get',
    S = 'set'
  };
  export var KEY_PREFIX_INJECT: '__$$_' = '__$$_'
  export var KEY_STRICT: 'strict' = 'strict'
  export var KEY_CONSTRUCTOR: 'constructor' = 'constructor';
  export var KEY_EXTEND: 'extend' = 'extend';
  export var KEY_FILTER: 'filter' = 'filter';
  export var KEY_PROPERTY: 'property' = 'property';
  export var KEY_PROTOTYPE: 'prototype' = 'prototype'
  export var KEY_NUM: 'number' = 'number';
  export var KEY_STR: 'string' = 'string';
  export var KEY_BOOL: 'boolean' = 'boolean';
  export var KEY_FUNC: 'function' = 'function';
  export var KEY_OBJ: 'object' = 'object';
  export var KEY_VAL: 'value' = DefPropDec$$.V;
  export var ARRAY = Array;
  export var IS_ARR = ARRAY.isArray;
  export var FUNCTION = Function;
  export var CALLER = FUNCTION.call
  export var Num = Number;
  export var REGEXP = RegExp;
  // tslint:disable-next-line: variable-name
  export var OBJECT = Object;
  export var KEY_DESIGN_TYPE: "design:type" = "design:type";
  export var KEY_DESIGN_RETURNTYPE: "design:returntype" = "design:returntype"; 
  export var KEY_DESIGN_PARAMTYPES: "design:paramtypes" = "design:paramtypes";
  export var EMPTY_OBJECT = OBJECT.seal(OBJECT.create(null));
	/**
	 * ```ts
	 * Object.keys
	 * ```
	 * */
  export var OBJ_KEYS = OBJECT.keys;
	/** 
	 * ```ts
	 * Object.freeze
	 * ``` 
	 * */
  export var OBJ_FREEZE = OBJECT.freeze;
	/**
	 * ```ts
	 * Object.assign
	 * ```
	 * */
  export var OBJ_ASSIGN = OBJECT.assign;
	/**
	 * ```ts
	 * Object.getOwnPropertyDescriptors
	 * ```
	 * */
  export var OBJ_getOwnPropertyDescriptors$ = OBJECT.getOwnPropertyDescriptors;
	/**
	 * ```ts
	 * Object.defineProperty
	 * ```
	 * */
  export var OBJ_defineProperty$ = OBJECT.defineProperty;
	/**
	 * ```ts
	 * Object.defineProperties
	 * ```
	 * */
  export var OBJ_defineProperties$ = OBJECT.defineProperties;

  const { V, E, C, W } = Constant$.DefPropDec$$
	/**
	 * ```ts
	 * Object.defineProperty
   * writable: true,
   * configurable: true,
   * enumerable: true,
   * value: param
	 * ```
	 * */
  export var OBJ_definePropertyNormal$ = function (target: any, key: Exclude<keyof typeof target, number>, value: any) {
    return OBJECT.defineProperty(target, key, {
      [V]: value, [W]: true, [C]: true, [E]: true
    })
  };
	/**
	 * ```ts
	 * Object.getPrototypeOf
	 * ```
	 * */
  export var OBJ_getPrototypeOf$ = OBJECT.getPrototypeOf;
	/**
	 * ```ts
	 * Object.getOwnPropertyDescriptor
	 * ```
	 * */
  export var OBJ_getOwnPropertyDescriptor$ = OBJECT.getOwnPropertyDescriptor;
	/**
	 * ```ts
	 * Object.getOwnPropertyNames
	 * ```
	 * */

  export var OBJ_getOwnPropertyNames$ = OBJECT.getOwnPropertyNames;
  export var PROMISE = Promise;
  export var CREATE_NEW = Reflect.construct

  export function CREATE_PROMISE<T>(executor: (resolve: (value?: T | PromiseLike<T>) => void, reject: (reason?: any) => void) => void): Promise<T> {
    return CREATE_NEW(PROMISE, [executor])
  }
  export function EMPTY_PROMISE<T = any>(value: T) {
    return PROMISE.resolve(value)
  }
  export var PARSE_FLOAT = parseFloat
  export var PARSE_INT = parseInt


  export function BIND<T extends FunctionFactory.Base<[any, ...any[]], any>,
    TA,
    R extends T extends (thisArgs: any, ...other: infer K) => infer Re ? FunctionFactory.Base<K, Re> : Function
  >(Func: T, thisArgs: TA, ...args: any[]): R {
    // @ts-ignore
    return FUNCTION.bind.call(Func, thisArgs, ...args)
  }

  // tslint:disable-next-line: class-name
  // export type STATIC_BIND<
  //   T extends Function,
  //   Bindex extends any[]
  // > = {
    
  // }
  export function BindArg$$<
    T extends Function,
    A extends FunctionFactory.Arg0<T>
    >(Func: T, arg0: A): FunctionFactory.Shift1<T, A>
  export function BindArg$$<
    T extends Function,
    A extends FunctionFactory.Arg0<T>,
    B extends FunctionFactory.Arg1<T>,
    >(Func: T, arg0: A, arg1: B): FunctionFactory.Shift2<T, A, B>;
  export function BindArg$$<
    T extends Function,
    A extends FunctionFactory.Arg0<T>,
    B extends FunctionFactory.Arg1<T>,
    C extends FunctionFactory.Arg2<T>,
    >(Func: T, arg0: A, arg1: B, arg2: C): FunctionFactory.Shift3<T, A, B, C>;
  export function BindArg$$(Func: any, ...args: any[]): any {
    // @ts-ignore
    return BIND(Func, null, ...args)
  }

  export function CREATE_OBJECT_IS(value: any): (value: any) => boolean {
    return BindArg$$(OBJECT.is, value)
  }

  export interface FunctionArgsShifter<
    T extends Function,
    P extends T['prototype'],
    K extends keyof P
    > {
    <Args extends P[K] extends (arg: infer A) => any ? A : any>(source: P, arg: Args): P[K] extends (...args: any[]) => infer R ? R : any;
  }

  export function INSTANCE_BIND<
    T extends Function,
    P extends T['prototype'],
    K extends keyof P,
  >(instance: T, key: K): FunctionArgsShifter<T, P, K> {
    // @ts-ignore
    return BIND(CALLER, instance[KEY_PROTOTYPE][key])
  }
  export function PROTOTYPE_BIND<
    T extends Function,
    P extends T['prototype'],
    K extends keyof P,
    >(instance: T, key: K): P[K] {
    // @ts-ignore
    return BIND(instance[KEY_PROTOTYPE][key], instance[KEY_PROTOTYPE]);
  }

  /**
   * 原生的map循环改为函数调用
   * @param arr
   * @param callbackfn
   */
  export const APPLY = (fn: any, args: any[], thisArgs?: any) => fn.apply(thisArgs, args) as {
    <TArgs extends any[], TResult = any, T = any>(source: Function, args: TArgs, thisArgs?: T): TResult
  }
  /**
   * 原生的map循环改为函数调用
   * @param arr
   * @param callbackfn
   */
  export const MAP = INSTANCE_BIND(ARRAY, 'map') as {
    <T extends any, R = T>(arr: T[], callbackfn: ArrayIterator<T, R>, initialValue?: any[]): R[]
  }

  export function DEFINED_ESMODULE(_exports: any, desc: any) {
    return Object.defineProperty(_exports, "__esModule", desc)
  }
  // MAP()
  /**
   * 原生的Array.prototype.some改为函数调用
   * @param arr
   * @param callbackfn
   * @param thisArg
   */
  export const SOME = INSTANCE_BIND(ARRAY, 'some') as {
    <T extends any>(arr: T[], callbackfn: ArrayIterator<T, boolean>, thisArg?: any): boolean
  }
  /**
   * 原生的for循环改为函数调用
   * @param arr
   * @param callbackfn
   */
  export const FOR_EACH = INSTANCE_BIND(ARRAY, 'forEach') as {
    <T extends any>(arr: T[], callbackfn: ArrayIterator<T, any>): void;
  };
  /**
   * 原生数组的concat改为函数调用
   * @param arr
   * @param callbackfn
   */
  export const ARR_CONCAT = PROTOTYPE_BIND(ARRAY, 'concat') as {
    <T extends any, A extends any = T>(arr: T[], ...next: A[][]): (A | T)[];
    <T extends any, A extends any = T>(arr: T[], ...next: A[]): (A | T)[];
  };
  /**
   * 原生数组的silce改为函数调用
   * @param arr
   * @param callbackfn
   */
  export const ARR_SLICE = INSTANCE_BIND(ARRAY, 'slice') as {
    <T extends any>(arr: T[], startIndex?: number, endIndex?: number): T[];
    // tslint:disable-next-line: unified-signatures
    <T extends any>(arr: IArguments, startIndex?: number, endIndex?: number): T[]; 
  }
  
  /**
   * 原生的for循环改为函数调用
   * @param arr
   * @param callbackfn
   */
  export const FILTER = INSTANCE_BIND(ARRAY, 'filter') as {
    <T extends any>(arr: T[], callbackfn: ArrayIterator<T, boolean>): T[];
  }


  /**
   * 原生的reduce改为函数调用
   * @param arr
   * @param callbackfn
   * @param initialValue
   * @example
   * REDUCE([1, 2, 3, 4], (target: any, i: number) => ({ ...target, [i]: i }), {})
   */
  export const REDUCE = INSTANCE_BIND(ARRAY, 'reduce') as {
    <T extends any, R = T>(
      arr: T[],
      callbackfn: (previousValue: R, currentValue: T, currentIndex: number, array: T[]) => R,
      initialValue?: R
    ): R
  }
  /**
   * 原生的Array<T>.push改为函数调用
   * @param arr 目标数组
   * @param elements 新元素
   * @example
   * ARR_PUSH([1, 2, 3, 4], 5, 6, 7)
   */
  export const ARR_PUSH = INSTANCE_BIND(ARRAY, 'push') as {
    <T extends any, E = T>(arr: T[], ...elements: E[]): number
  }

  /**
   * 原生的push改为函数调用
   * @param arr 目标数组
   * @param elements 新元素
   * @example
   * REGEXP_TEST(/123/, '123') // => true
   */
  export const REGEXP_TEST = INSTANCE_BIND(REGEXP, 'test') as {
    (reg: RegExp, str: string): boolean
  }

  export const delay$$ = BIND(setTimeout, null);

  /**
   *
   * @param arr
   * @param callbackfn
   */
  export function MAP$$<T, R>(arr: T[], callbackfn: ArrayIterator<T, R>): R[]
  export function MAP$$<T, R>(arr: T[], callbackfn: ArrayIterator<T, R>, initialValue = [], i: number = 0, length = arr.length - 1) {
    initialValue[i] = callbackfn(arr[i], i, arr)
    // @ts-ignore
    return i < length ? MAP$$(arr, callbackfn, initialValue, i + 1, length) : initialValue
  }


  /**
   *
   * @param arr
   * @param callbackfn
   */
  export function FOR_EACH$$<T>(arr: T[], callbackfn: ArrayIterator<T, any>): true {
    var tmp = { i: 0 }
    var length = arr.length - 1
    try {
      return FOR_EACH_P$$(arr, callbackfn, tmp, length)
    } catch (error) {
      while (length - tmp.i++) callbackfn(arr[tmp.i], tmp.i, arr)
      return true
    }
  }
  function FOR_EACH_P$$<T>(arr: T[], callbackfn: ArrayIterator<T, any>, tmp: { i: number }, length: number) {
    callbackfn(arr[tmp.i], tmp.i, arr)
    // console.trace('test')
    // @ts-ignore
    return tmp.i++ < length ? FOR_EACH_P$$(arr, callbackfn, tmp, length) : true
  }


  //
  export var Key_useDeprecatedSynchronousErrorHandling$$ = 'useDeprecatedSynchronousErrorHandling'
}

export type ConstructorType<T, Args extends any[] = [any?, any?, ...any[]]> = {
  new(...args: Args): InstanceType<new (...args: any) => T>;
};
export type TKey = string | number | symbol;
export type IKeyValueMap<V = any, K extends TKey = TKey> = Record<K, V>
declare global {
  // var K$: typeof Constant$;
  type ConstructorType<T, Args extends any[] = [any?, any?, ...any[]]> = {
    new(...args: Args): InstanceType<new (...args: any) => T>;
  };
  type TKey = string | number | symbol;
  type IKeyValueMap<V = any, K extends TKey = TKey> = Record<K, V>
}

// @ts-ignore
// tslint:disable-next-line: no-eval
// K$ = eval('Constant$')
