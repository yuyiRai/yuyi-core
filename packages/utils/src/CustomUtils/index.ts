/**
 * @module CustomUtils
 */

import { zipEmptyData } from './zipEmptyData';
export * from './pipe';
import { uuid } from './uuid';
import { downloadFile } from './downloadFile';
import { createObjectKey } from './createObjectKey';
import { getListDifferent } from './getListDifferent';
import { connectTo } from './connectTo';
// import { jsGetAge } from './jsGetAge';
import { likeArray } from './likeArray';
import { getRealLength } from './getRealLength'
import { MultipleSplit } from './MultipleSplit'
export * from './changeKeys';
export * from './waitingPromise';
export * from './autobind'

export {
  uuid,
  downloadFile,
  zipEmptyData,
  createObjectKey,
  getListDifferent,
  getRealLength,
  connectTo,
  // jsGetAge,
  likeArray,
  MultipleSplit
}
// export * from './getTimeProfile'
