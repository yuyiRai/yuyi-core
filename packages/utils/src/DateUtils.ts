import { isDate } from "./LodashExtra";
import momentType from 'moment'
import moment from "dayjs";
import { expect$ } from "./TypeLib";
import { IsAny, IsUnknown } from './TsUtils';
// export enum EDateFormatter {
//   /**
//    * 日期
//    */
//   date = 'YYYY-MM-DD',
//   /**
//    * 日期时分秒
//    */
//   dateTime = 'YYYY-MM-DD HH:mm:ss'
// }
// export type DateFormatter = EDateFormatter | string

export type InputType = IsUnknown<moment.ConfigType, momentType.MomentInput, moment.ConfigType>
export class DateUtils {
  public static format(value: any, formatter: EDateFormatter = EDateFormatter.dateTime): string {
    return toDateString(value, formatter);
  }

  public static hoursDiff(targetDate: InputType, currentDate?: InputType) {
    return Math.round((moment(targetDate as moment.ConfigType).diff(moment(currentDate as moment.ConfigType)) / 1000 / 60 / 60))
  }

  public static getDateRange(days: number) {
    const end = new Date();
    const start = new Date();
    start.setTime(start.getTime() - 3600 * 1000 * 24 * days);
    return [start, end];
  }
}
/**
 * 常用formatter模板
 */
export enum EDateFormatter {
  /**
   * 日期
   */
  date,
  /**
   * 日期时分秒
   */
  dateTime
}
const values = ['YYYY-MM-DD', 'YYYY-MM-DD HH:mm:ss']
export type DateFormatter = EDateFormatter | string;
export function toDateString(value: any, formatter: EDateFormatter = EDateFormatter.dateTime): string {
  const format = values[formatter]
  if (isDate(value)) {
    return moment(value).format(format);
  } else if (expect$.isNotEmptyString(value)) {
    const date = moment(value);
    if (date.isValid()) {
      return date.format(format);
    }
  }
  return moment().format(format);
}
