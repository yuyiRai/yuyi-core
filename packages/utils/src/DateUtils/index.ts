import moment from "dayjs";
import momentType from 'moment';
import { isDate, isNum } from "../LodashExtra";
import { IsUnknown } from '../TsUtils';
import { expect$ } from "../TypeLib";

/**
 * 常用formatter模板
 */
export enum EDateFormatter {
  /**
   * 日期
   */
  date,
  /**
   * 日期时分秒
   */
  dateTime
}
const EDateFormatterValues = ['YYYY-MM-DD', 'YYYY-MM-DD HH:mm:ss'];
export type DateFormatter = EDateFormatter | string;
export type TDate = IsUnknown<
  moment.ConfigType,
  momentType.MomentInput,
  moment.ConfigType
  >;
export type TMoment = IsUnknown<
  moment.Dayjs,
  typeof momentType,
  moment.Dayjs
>


/**
 * 
 * 比较生日和当前日期取得年龄
 * @param birthday - 生日
 * @param offset - 偏移量
 * @param current - 替换当前日期
 */
export function getAge(birthday: TDate, offset?: number): number;
export function getAge(birthday: TDate, current: TDate): number;
export function getAge(birthday: TDate, current: TDate | number = new Date()) {
  const sBirthday = moment(birthday);
  const sNow = isNum(current) ? getDay(current, false) : moment(current);
  let birthYear = sBirthday.year();
  let nowYear = sNow.year();
  if (nowYear === birthYear) return 0; // 同年 则为0岁
  let birthMonth = sBirthday.month() + 1;
  let birthDay = sBirthday.date();
  let nowMonth = sNow.month() + 1;
  let nowDay = sNow.date();
  let ageDiff = nowYear - birthYear; // 年之差
  if (ageDiff > 0) {
    if (nowMonth === birthMonth) {
      let dayDiff = nowDay - birthDay; // 日之差
      return dayDiff < 0 ? (ageDiff - 1) : ageDiff;
    }
    // tslint:disable-next-line: radix
    let monthDiff = nowMonth - birthMonth; // 月之差
    return monthDiff < 0 ? (ageDiff - 1) : ageDiff;
  }
  return -1; // 返回-1 表示出生日期输入错误 晚于今天
}

export function format(value: any, formatter: EDateFormatter = EDateFormatter.dateTime): string {
  return toDateString(value, formatter);
}

/**
 * 取得日期
 * @param offset 日期偏移，大于0返回未来n天，小于0表示过去n天
 */
export function getDay(format?: DateFormatter | boolean): TMoment
export function getDay(offset: number, format?: false): TMoment
export function getDay(offset: number, format?: true | DateFormatter): string;
export function getDay(arg1: any, arg2: any = false) {
  const mode = isNum(arg1)
  const offset: number = mode ? Math.round(arg1) : 0;
  const format: boolean | DateFormatter = mode ? arg2 : arg1;
  const day = moment();
  offset !== 0 && day.set('date', day.get('date') + offset);
  return format
    ? day.format(expect$.isString.filter(format, EDateFormatterValues[format as EDateFormatter], 'YYYY-MM-DD HH:mm:ss'))
    : day
}

/**
 * 返回昨天
 * @param offset 天数偏移，传入小于等于0时也会返回昨天
 */
export function getYesterday(offset: number = 0, format?: boolean | DateFormatter) {
  return getDay(-1 - Math.max(offset, 0), format as any)
}

/**
 * 返回明天
 * @param offset 天数偏移，传入小于等于0时也会返回明天
 */
export function getFuture(offset: number = 0) {
  return getDay(1 + Math.max(offset, 0));
}

export function hoursDiff(targetDate: TDate, currentDate ?: TDate) {
  return (moment(targetDate as moment.ConfigType).diff(moment(currentDate as moment.ConfigType)) / 1000 / 60 / 60);
}

export function getDateRange(days: number, end = new Date()) {
  const start = new Date(end);
  start.setTime(start.getTime() - 3600 * 1000 * 24 * days);
  return [start, end];
}

export function toDateString(value: any, formatter: EDateFormatter = EDateFormatter.dateTime): string {
  const format = EDateFormatterValues[formatter]
  if (isDate(value)) {
    return moment(value).format(format);
  } else if (expect$.isNotEmptyString(value)) {
    const date = moment(value);
    if (date.isValid()) {
      return date.format(format);
    }
  }
  return moment().format(format);
}
