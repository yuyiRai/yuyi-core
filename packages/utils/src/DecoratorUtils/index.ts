export * from './createDecoratorBase';
export * from './generateNamespace';
export * from './logger';
export * from './Trasaction';
export * from './getConstructor';

