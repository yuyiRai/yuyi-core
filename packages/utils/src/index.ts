/**
 * @author yuyi
 * @packageDocumentation
 */
import './Constransts'
// import { oc } from 'ts-optchain';
// import { ArrayList } from 'typescriptcollectionsframework'
// Import here Polyfills if needed. Recommended core-js (npm i -D core-js)
// import "core-js/fn/array.find"
// ...
// import { LodashExtraUtils } from './LodashExtra'
// import '@yuyi919/env'

// const A = Type
// export const Constant$$ = { ...Constant$ }
// export const AAA = { a: { b: { c: 3 } } }
// export const AAAA = oc(AAA).a.b.c

// export const A = new ArrayList(undefined);

export * from './CustomUtils'
export * from './OptionsUtils'
export * from './TimeBuffer'
export * from './LodashExtra'
export * from './FormGirdUtils'
export * from './TypeLib'
export * from './HttpBox'
export * from './PropertyUtils'
export * from './commonUtils'
export * from './TestUtils'
export * from './ParseUtils'
export * from './MobxUtils'
export * from './EventEmitter'
export * from './WasmLoader'

// export * from './Utils'
