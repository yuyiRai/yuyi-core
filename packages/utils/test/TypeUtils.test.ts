import { EventEmitter, expect$, FilterFunction, isEmptyObject, isFunction, isNotEmptyObject, isNotEmptyValue } from '../src'
import { Constant$ } from '../src/Constransts'

/**
 * 类型判断工具测试
 */
describe('TypeUtils test', () => {
  const expectMap = new Map()
  expectMap.set({}, true)
  expectMap.set([], true)
  expectMap.set(true, true)
  expectMap.set(0, true)
  expectMap.set('', false)
  expectMap.set('N', true)
  expectMap.set(null, false)
  expectMap.set(undefined, false)
  expectMap.set(NaN, false)
  expectMap.forEach((r, v) => {
    it('isNotEmptyValue:' + JSON.stringify(v) + '=' + r, () => {
      expect(isNotEmptyValue(v)).toBe(r)
    })
  })

  // tslint:disable-next-line: no-empty
  function TestFunction() { }
  const testGroupA = [NaN, null, undefined, 1, {}, [null], [1], 1, 20, '1,2,3']
  const keys: (keyof typeof expect$)[] = Object.keys(expect$) as any
  Constant$.FOR_EACH(keys, (key) => {
    if (key !== 'filter' && key !== 'pure') {
      it(`测试 expect$.filters - ${key}`, () => {
        expect((expect$[key].filter as FilterFunction<any>)(...testGroupA)).toMatchSnapshot()
      })
    }
  })
  it('isEmpty etc.', () => {
    expect(expect$.isEmptyObject({})).toBe(true)
    expect(isEmptyObject({ test: 1 })).toBe(false)
    expect(isEmptyObject(Object.create(null))).toBe(true)
    expect(isEmptyObject(Function)).toBe(false)
    expect(isEmptyObject(TestFunction)).toBe(false)

    expect(expect$.filter.isNotEmptyObject<'string'>(...testGroupA)).toBe(undefined)
    expect(expect$.filter.isNotEmptyData<EventEmitter>(...testGroupA)).toEqual(1)
    expect(expect$.filter.isEmptyObject(...testGroupA)).toEqual({})
    expect(expect$.isNotNil.filter<string>(...testGroupA)).toEqual(NaN)
    expect(expect$.isNil.filter<string>(...testGroupA)).toEqual(null)
    expect(expect$.isEventEmitter.filter<'a'>(...testGroupA)).toBe(undefined)
    expect(expect$.isString.filter<'a'>(...testGroupA)).toBe('1,2,3')
    expect(expect$.isBoolean.filter(...testGroupA)).toBe(undefined)

    expect(expect$.isArray.filter<'str'>(...testGroupA)).toEqual([null])
    expect(expect$.isNotEmptyArray.filter(...testGroupA)).toEqual([null])
    expect(expect$.isNotEmptyArrayStrict.filter(...testGroupA)).toEqual([1])

    expect(isNotEmptyObject({})).toBe(false)
    expect(isNotEmptyObject({ test: 1 })).toBe(true)
    
  })


  it('native instanceof', () => {
    const r = TestFunction instanceof Function
    const r2 = isFunction(TestFunction)

    expect(r).toBeTruthy()
    expect(r2).toBeTruthy()
  })
})
