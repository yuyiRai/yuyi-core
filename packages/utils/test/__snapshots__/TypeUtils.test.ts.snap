// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeUtils test 测试 expect$.filters - isArray 1`] = `
Array [
  null,
]
`;

exports[`TypeUtils test 测试 expect$.filters - isArrayLike 1`] = `
Array [
  null,
]
`;

exports[`TypeUtils test 测试 expect$.filters - isBoolean 1`] = `undefined`;

exports[`TypeUtils test 测试 expect$.filters - isBooleanOrNumber 1`] = `1`;

exports[`TypeUtils test 测试 expect$.filters - isDate 1`] = `undefined`;

exports[`TypeUtils test 测试 expect$.filters - isEmptyArray 1`] = `undefined`;

exports[`TypeUtils test 测试 expect$.filters - isEmptyArrayStrict 1`] = `
Array [
  null,
]
`;

exports[`TypeUtils test 测试 expect$.filters - isEmptyData 1`] = `NaN`;

exports[`TypeUtils test 测试 expect$.filters - isEmptyObject 1`] = `Object {}`;

exports[`TypeUtils test 测试 expect$.filters - isEmptyValue 1`] = `NaN`;

exports[`TypeUtils test 测试 expect$.filters - isEventEmitter 1`] = `undefined`;

exports[`TypeUtils test 测试 expect$.filters - isFunction 1`] = `undefined`;

exports[`TypeUtils test 测试 expect$.filters - isNaN 1`] = `NaN`;

exports[`TypeUtils test 测试 expect$.filters - isNil 1`] = `null`;

exports[`TypeUtils test 测试 expect$.filters - isNotEmptyArray 1`] = `
Array [
  null,
]
`;

exports[`TypeUtils test 测试 expect$.filters - isNotEmptyArrayStrict 1`] = `
Array [
  1,
]
`;

exports[`TypeUtils test 测试 expect$.filters - isNotEmptyData 1`] = `1`;

exports[`TypeUtils test 测试 expect$.filters - isNotEmptyObject 1`] = `undefined`;

exports[`TypeUtils test 测试 expect$.filters - isNotEmptyString 1`] = `"1,2,3"`;

exports[`TypeUtils test 测试 expect$.filters - isNotEmptyValue 1`] = `1`;

exports[`TypeUtils test 测试 expect$.filters - isNotNil 1`] = `NaN`;

exports[`TypeUtils test 测试 expect$.filters - isNumber 1`] = `1`;

exports[`TypeUtils test 测试 expect$.filters - isObject 1`] = `Object {}`;

exports[`TypeUtils test 测试 expect$.filters - isPlainObject 1`] = `Object {}`;

exports[`TypeUtils test 测试 expect$.filters - isPureObj 1`] = `Object {}`;

exports[`TypeUtils test 测试 expect$.filters - isString 1`] = `"1,2,3"`;
